File: frontend/src/lib/api/telemetry.ts
import { API_BASE } from "./client";

export interface ClientErrorReport {
  message: string;
  stack?: string;
  component_stack?: string;
  url?: string;
}

/**
 * Best-effort client-error reporting. Fire-and-forget: never throws, never
 * blocks the UI, and silently no-ops if the backend endpoint is unavailable.
 * The ErrorBoundary always logs to the console regardless.
 */
export function reportClientError(report: ClientErrorReport): void {
  try {
    void fetch(`${API_BASE}/telemetry/client-error`, {
      method: "POST",
      headers: { "content-type": "application/json" },
      credentials: "include",
      body: JSON.stringify({ ...report, url: report.url ?? window.location.href }),
      keepalive: true,
    }).catch(() => {
      /* ignore — reporting is non-critical */
    });
  } catch {
    /* ignore */
  }
}

File: backend/app/schemas/telemetry.py
"""Schemas for lightweight client-side telemetry."""

from __future__ import annotations

from pydantic import BaseModel, Field


class ClientErrorIn(BaseModel):
    """A client-side (frontend) error reported by the ErrorBoundary."""

    message: str = Field(..., max_length=2000)
    stack: str | None = Field(default=None, max_length=8000)
    component_stack: str | None = Field(default=None, max_length=8000)
    url: str | None = Field(default=None, max_length=1000)
